datajunction-server/datajunction_server/sql/dag.py

```diff
@@ -1,22 +1,244 @@
+            )
+            .unique()
+            .all()
+        )
+
+    # Otherwise return the dimension attributes, which include both the dimension
+    # attributes on the dimension nodes in the DAG as well as the local dimension
+    # attributes on the initial node
+    group_concat = (
+        func.group_concat
+        if session.bind.dialect.name in ("sqlite",)
+        else func.string_agg
+    )
+    final_query = (
+        select(
+            paths.c.node_name,
+            paths.c.node_display_name,
+            column.name,
+            column.type,
+            group_concat(AttributeType.name, ",").label(
+                "column_attribute_type_name",
+            ),
+            paths.c.join_path,
+        )
+        .select_from(paths)
+        .join(NodeColumns, NodeColumns.node_id == paths.c.node_revision_id)
+        .join(column, NodeColumns.column_id == column.id)
+        .join(ColumnAttribute, column.id == ColumnAttribute.column_id, isouter=True)
+        .join(
+            AttributeType,
+            ColumnAttribute.attribute_type_id == AttributeType.id,
+            isouter=True,
+        )
+        .group_by(
+            paths.c.node_name,
+            paths.c.node_display_name,
+            column.name,
+            column.type,
+            paths.c.join_path,
+        )
+        .union_all(
+            select(
+                NodeRevision.name,
+                NodeRevision.display_name,
+                Column.name,
+                Column.type,
+                group_concat(AttributeType.name, ",").label(
+                    "column_attribute_type_name",
+                ),
+                literal("").label("join_path"),
+            )
+            .select_from(NodeRevision)
+            .join(NodeColumns, NodeColumns.node_id == NodeRevision.id)
+            .join(Column, NodeColumns.column_id == Column.id)
+            .join(
+                ColumnAttribute,
+                Column.id == ColumnAttribute.column_id,
+                isouter=True,
+            )
+            .join(
+                AttributeType,
+                ColumnAttribute.attribute_type_id == AttributeType.id,
+                isouter=True,
+            )
+            .group_by(
+                NodeRevision.name,
+                NodeRevision.display_name,
+                Column.name,
+                Column.type,
+                "join_path",
+            )
+            .where(NodeRevision.id == node_revision.id),
+        )
+    )
+
+    return sorted(
+        [
+            DimensionAttributeOutput(
+                name=f"{row[0]}.{row[2]}",
+                node_name=row[0],
+                node_display_name=row[1],
+                is_primary_key=row[4] == "primary_key",
+                type=str(row[3]),
+                path=row[5].split(",")[:-1] if row[5] else [],
+            )
+            for row in session.exec(final_query).all()
+            if row[5] != ""
```

Comment:
Could you please add some comments here. At least to remind what `row[5]` is and why are you looking for empty string (instead of None).

Reply:
Yep, I added some comments